### PR TITLE
feat(tailwind): package rewrite

### PR DESCRIPTION
**Pull request by lordelogos**
### Description
This PR is a rewrite of the `<Tailwind>` component.
It contains fixes for:
- #890 
- #891 
- #943 

It also removes the need for `html-react-parser`

### CODE REVIEW

File: packages/tailwind/src/tailwind.tsx [hunk 1]
```diff
@@ -1,179 +1,142 @@
...
-  const newChildren = React.Children.toArray(children);
-
-  const fullHTML = renderToStaticMarkup(<>{newChildren}</>);
+  if (element.props.className) {
+    const convertedStyles: string[] = [];
+    const responsiveStyles: string[] = [];
+    const classNames = element.props.className.split(" ");
+
+    const customClassNames = classNames.filter((className: string) => {
+      if (twi(className, { ignoreMediaQueries: true })) {
```
Comment (vcapretz) [inline packages/tailwind/src/tailwind.tsx:27]:
```suggestion
      const tailwindClassName = twi(className, { ignoreMediaQueries: true })
      if (tailwindClassName) {
```

so we also reuse this in line 28

File: packages/tailwind/src/tailwind.tsx [hunk 2]
```diff
@@ -1,179 +1,142 @@
...
+  });
+
+  if (!childrenWithInlineStyles) return <>{children}</>;
+
+  const fullHTML = renderToStaticMarkup(<>{childrenWithInlineStyles}</>);
+
+  const hasResponsiveStyles = /@media[^{]+\{(?<content>[\s\S]+?)\}\s*\}/gm.test(
+    headStyles.join(" "),
   );
+  const hasHTML = /<html[^>]*>/gm.test(fullHTML);
```
Comment (vcapretz) [inline packages/tailwind/src/tailwind.tsx:120]: can this and `hasHead` be done in one pass so we don't test the full html twice?

File: packages/tailwind/src/tailwind.tsx [hunk 3]
```diff
@@ -1,179 +1,142 @@
...
     },
-    {} as Record<string, string>,
   );
-  return cssMap;
-}
+
+  return <>{childrenWithInlineAndResponsiveStyles}</>;
+};
+
+Tailwind.displayName = "Tailwind";
```
Comment (vcapretz) [inline packages/tailwind/src/tailwind.tsx:142]: we can remove this now that the component doesn't have `forwardRef` anymore 🙏

**Review by vcapretz — APPROVED** [review]

**Comment by maccman** [comment]
whoot whooot!!! Amazing!!

**Comment by AntonioJABV** [comment]
Hi there,

I was wondering if you have an estimated timeline for when the TailwindCSS fix from the recently merged PR #953 will be included in the next stable release of the package?

I'm looking forward to using the fix in my project.

Please let me know, thanks!

File: packages/tailwind/src/tailwind.tsx [hunk 4]
```diff
@@ -1,104 +1,162 @@
...
+import type { TailwindConfig, twi as TwiType } from "tw-to-css";
+import { tailwindToCSS } from "tw-to-css";
+import { cssToJsxStyle } from "./utils/css-to-jsx-style";
 
 export interface TailwindProps {
   children: React.ReactNode;
   config?: TailwindConfig;
 }
 
+function processElement(
```
Comment (roderickhsiao) [inline packages/tailwind/src/tailwind.tsx:14]: it seems like the new process method behaves differently then the previous parser. Haven't debug deeper but after upgrade to the latest version, some Tailwind class no longer parsed.

Will try to provide a code sandbox to reproduce the issue

| **Before** | **After** |
| -- | -- |
| Screenshot 2023-10-31 at 11 53 31 PM | Screenshot 2023-10-31 at 11 53 22 PM |

Comment (grahamplace) [reply packages/tailwind/src/tailwind.tsx:14]: +1, I tried to bump `react-email` versions last night so I could upgrade my project to Next 14, and that appears to have broken all my emails, which are all wrapped with a `<Tailwind>` wrapper (padding, background, text color set with tailwind classes are the obvious things not working, not sure exactly why yet)

Agree that `"@react-email/tailwind": "0.0.9",` appears to be the last working version

Comment (grahamplace) [reply packages/tailwind/src/tailwind.tsx:14]: Although, I should say I'm not sure exactly what is/isn't working. E.g. the following, running with Next14 is producing the same broken Tailwind email components for me
```
    "@react-email/components": "0.0.11",
    "@react-email/render": "0.0.9",
    "@react-email/tailwind": "0.0.9",
```

**Pull request merged**